UpdateSubscriptionCidrs/handler.ts

```diff
@@ -0,0 +1,126 @@
+import { ContextMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/context_middleware";
+import { RequiredParamMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/required_param";
+import { withRequestMiddlewares } from "@pagopa/io-functions-commons/dist/src/utils/request_middleware";
+import { wrapRequestHandler } from "@pagopa/ts-commons/lib/request_middleware";
+import {
+  IResponseErrorInternal,
+  IResponseErrorNotFound,
+  IResponseSuccessJson,
+  ResponseErrorInternal,
+  ResponseSuccessJson
+} from "@pagopa/ts-commons/lib/responses";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+
+import { RequiredBodyPayloadMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/middlewares/required_body_payload";
+import {
+  IResponseErrorQuery,
+  ResponseErrorQuery
+} from "@pagopa/io-functions-commons/dist/src/utils/response";
+import { toAuthorizedCIDRs } from "@pagopa/io-functions-commons/dist/src/models/service";
+import { SubscriptionCIDRsModel } from "@pagopa/io-functions-commons/dist/src/models/subscription_cidrs";
+import {
+  getApiClient,
+  IAzureApimConfig,
+  IServicePrincipalCreds
+} from "../utils/apim";
+import { CIDRsPayload } from "../generated/definitions/CIDRsPayload";
+
+type IUpdateSubscriptionCidrsHandler = (
+  context: Context,
+  auth: IAzureApiAuthorization,
+  subscriptionid: NonEmptyString,
+  cidrsPayload: CIDRsPayload
+) => Promise<
+  | IResponseSuccessJson<CIDRsPayload>
+  | IResponseErrorInternal
+  | IResponseErrorNotFound
+  | IResponseErrorQuery
+>;
+
+// eslint-disable-next-line prefer-arrow/prefer-arrow-functions
+export function UpdateSubscriptionCidrsHandler(
+  servicePrincipalCreds: IServicePrincipalCreds,
+  azureApimConfig: IAzureApimConfig,
+  subscriptionCIDRsModel: SubscriptionCIDRsModel
+): IUpdateSubscriptionCidrsHandler {
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+  return async (context, _, subscriptionId, cidrs) => {
+    const errorOrApiManagementClient = await getApiClient(
+      servicePrincipalCreds,
+      azureApimConfig.subscriptionId
+    )();
+    if (E.isLeft(errorOrApiManagementClient)) {
+      return ResponseErrorInternal("Error trying to get Api Management Client");
+    }
+
+    const apiClient = errorOrApiManagementClient.right;
+    const errorOrSubscriptionResponse = await TE.tryCatch(
+      () =>
+        apiClient.subscription.get(
+          azureApimConfig.apimResourceGroup,
+          azureApimConfig.apim,
+          subscriptionId
+        ),
+      E.toError
+    )();
+    if (E.isLeft(errorOrSubscriptionResponse)) {
+      return ResponseErrorInternal("Error trying to get APIM Subscription");
+    }
+
+    const errorOrMaybeUpdatedSubscriptionCIDRs = await subscriptionCIDRsModel.upsert(
+      {
+        cidrs: toAuthorizedCIDRs(Array.from(cidrs)),
+        kind: "INewSubscriptionCIDRs",
+        subscriptionId
+      }
+    )();
```

Comment:
I really like the idea of the third option, because it would cover both streams (`manage` and `use`).
If we decide to change from the current situation _(option 1)_, I would say start with option 2, and then extend the behavior to reach option 3 as well, if needed and helpful.
Let's take a further step in the discussion about it.

Reply:
As discussed, we leave the current option 1 and add some explanation on API

Reply:
e62096f